jobbergate-api/jobbergate_api/apps/services.py

```diff
@@ -1,140 +1,464 @@
+
+        self.session.add(instance)
+        # I think this flush is not necessary
+        await self.session.flush()
+        await self.session.refresh(instance)
+        return instance
 
     async def count(self) -> int:
-        """Count the number of entries on the database."""
-        statement = select(func.count()).select_from(self.base_table)
-        return (await self.db_session.execute(statement)).scalar_one()
-
-    async def get(self, id_or_identifier: int | str) -> Base:
-        """Get an entry by id."""
-        query = select(self.base_table)
-        query = self._locate_by_id_or_identifier(id_or_identifier, query)
-        result = await self.db_session.execute(query)
-        row = result.scalar_one_or_none()
-        if row is None:
-            self._handle_not_found(id_or_identifier)
-        return row
-
-    async def delete(self, id_or_identifier: int | str) -> None:
-        """Delete an entry by id."""
-        query = delete(self.base_table)
-        query = self._locate_by_id_or_identifier(id_or_identifier, query)
-        result = await self.db_session.execute(query)
-        if result.rowcount != 1:
-            self._handle_not_found(id_or_identifier)
-
-    async def update(self, id_or_identifier: int | str, **incoming_data) -> Base:
-        """Update an entry by id."""
-        query = update(self.base_table).returning(self.base_table)
-        query = self._locate_by_id_or_identifier(id_or_identifier, query)
-        query = query.values(**incoming_data)
-        result = await self.db_session.execute(query)
-        row = result.scalar_one_or_none()
-        if row is None:
-            self._handle_not_found(id_or_identifier)
-        return row
-
-    async def list(
+        """
+        Count the number of rows in the table on the database.
+        """
+        result: Result = await self.session.execute(select(func.count(self.model_type.id)))
+        return result.scalar_one()
+
+    async def get(self, locator: Any) -> CrudModel:
+        """
+        Get a row by locator.
+
+        In almost all cases, the locator will just be an ``id`` value.
+        """
+        query = select(self.model_type).where(self.locate_where_clause(locator))
+        result: Result = await self.session.execute(query)
+        instance: CrudModel = enforce_defined(
+            result.scalar_one_or_none(),
+            f"{self.model_type.__tablename__} row not found by {locator}",
+            raise_exc_class=ServiceError,
+            raise_kwargs=dict(status_code=status.HTTP_404_NOT_FOUND),
+        )
+        await self.session.refresh(instance)
+        return instance
+
+    async def delete(self, locator: Any) -> None:
+        """
+        Delete a row by locator.
+
+        In almost all cases, the locator will just be an ``id`` value.
+        """
+        query = delete(self.model_type).returning(self.model_type).where(self.locate_where_clause(locator))
```

Comment:
It allows us to count the number of rows deleted so that we can ensure that the delete matched the row. Otherwise, we would need to do a separate get and then delete.